deployer/helm_upgrade_decision.py

```diff
@@ -0,0 +1,367 @@
+    Actions in order to generate parallel jobs in a matrix.
+    Note: "cluster folders" are those that contain a cluster.yaml file.
+    Args:
+        modified_dirpaths (list[path obj]): List of absolute paths to cluster folders
+            that contain EITHER a modified cluster.yaml OR a modified support.values.yaml
+            file
+        upgrade_all_clusters (bool, optional): If True, generates jobs to update the
+            support chart on all clusters. This is triggered when common config has been
+            modified in the support helm chart. Defaults to False.
+    Returns:
+        list[dict]: A list of dictionaries. Each dictionary contains: the name of a
+            cluster and the cloud provider that cluster runs on.
+    """
+    # Empty list to store the matrix definitions in
+    matrix_jobs = []
+
+    if upgrade_all_clusters:
+        print(
+            "Common config has been updated. Generating jobs to upgrade support chart on ALL clusters."
+        )
+
+        # Overwrite cluster_filepaths to contain paths to all clusters
+        if test_env == "test":
+            # We are running a test via pytest. We only want to focus on the cluster
+            # folders nested under the `tests/` folder.
+            modified_dirpaths = [
+                filepath.parent
+                for filepath in Path(os.getcwd()).glob("**/cluster.yaml")
+                if "tests/" in str(filepath)
+            ]
+        else:
+            # We are NOT running a test via pytest. We want to explicitly ignore the
+            # cluster folders nested under the `tests/` folder.
+            modified_dirpaths = [
+                filepath.parent
+                for filepath in Path(os.getcwd()).glob("**/cluster.yaml")
+                if "tests/" not in str(filepath)
+            ]
+
+    for cluster_filepath in modified_dirpaths:
+        # Read in the cluster.yaml file
+        with open(cluster_filepath.joinpath("cluster.yaml")) as f:
+            cluster_config = yaml.load(f)
+
+        # Generate a dictionary-style job entry for this cluster
+        cluster_info = {
+            "cluster_name": cluster_config.get("name", {}),
+            "provider": cluster_config.get("provider", {}),
+        }
+
+        # Double-check that support is defined for this cluster. If it is, append the
+        # job definition to the list of matrix jobs.
+        support_config = cluster_config.get("support", {})
+        if support_config:
+            matrix_jobs.append(cluster_info)
+        else:
+            print(f"No support defined for cluster: {cluster_config.get('name', {})}")
+
+    return matrix_jobs
+
+
+def update_github_env(hub_matrix_jobs, support_matrix_jobs):
+    """Update the GITHUB_ENV environment with two new variables describing the matrix
+    jobs that need to be run in order to update the support charts and hubs that have
+    been modified.
```

Comment:
_Original discussion:_ https://github.com/2i2c-org/infrastructure/pull/1093#discussion_r826233431

> [consideRatio](https://github.com/consideRatio) [21 hours ago](https://github.com/2i2c-org/infrastructure/pull/1093#discussion_r826233431)
> Can you rephrase this a bit to be clearer on how and why it's done? My understanding is that GITHUB_ENV is a path to a file, which if we write to, influences the environment as observed by a GitHub Workflow's Job's subsequent steps.
> 
> I'm also thinking perhaps we should fail loudly and early if we end up running functions that assume we are doing it as part of a GitHub workflow execution - but aren't.
> 
> From [docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables](https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables):
> 
> GITHUB_ACTIONS | Always set to true when GitHub Actions is running the workflow. You can use this variable to differentiate when tests are being run locally or by GitHub Actions.
>  
> Member
> Author
> @[sgibson91](https://github.com/sgibson91) sgibson91 [1 hour ago](https://github.com/2i2c-org/infrastructure/pull/1093#discussion_r827048183) • 
> So the decision is made to print rich tables if EITHER the --pretty-print argument is passed OR GITHUB_ENV is found in the environment:
> 
> [sgibson91/pilot-hubs@5b8d720/deployer/deploy_actions.py#L267-L272](https://github.com/sgibson91/pilot-hubs/blob/5b8d720c39a20fa1b09de4f026f7d172c5fe4eaa/deployer/deploy_actions.py#L267-L272)
> 
> I'm also thinking perhaps we should fail loudly and early if we end up running functions that assume we are doing it as part of a GitHub workflow execution - but aren't.
> 
> So we don't fail, but instead print out human-readable output instead of trying to write to the GITHUB_ENV file

Reply:
This was entirely about code comprehension and can be ignored as well.

The key point of confusion I needed clarify in the docstring was about GITHUB_ENV: a technical environment variable used in GitHub workflows. It includes the path to a file. If that file is appended with environment variables in a `HELLO=VALUE` format for example, they will be set in a subsequent step of the github workflow's job we are currently executing in.